### ISSUE TITLE
Add copy-constructor for Elem

### ISSUE THREAD

**Issue by loganharbour** [libMesh/libmesh#2797]
In order to future proof the copying of an Elem from another, a copy constructor should be added that can be updated accordingly. It will do all of the necessary copy spare for the nodes, which will be left nullptr.

There's already plenty of use cases for this that exist for testing.

**Comment by jwpeterson** [comment]
> There's already plenty of use cases for this that exist for testing.

Not sure what you mean? Are there cases of Elem copying where we do it "by hand" instead of using a copy constructor? 

I'm not saying we shouldn't add an Elem copy constructor, but just to keep in mind that hand-coded copy/move operations often become sources of bugs because people forget to update them whenever they add something new to the class. So, best to have a good reason for adding them in the first place (unless they can be defaulted of course).

**Comment by loganharbour** [comment]
This actually came as a suggestion from @roystgnr. 

If we get enough to rely on this copy constructor, there shouldn't be an issue with it being a source of bugs - it'll actually make it easier when making future changes to Elem, along with hopefully a more obvious failure I'd think.

The intention is to use it here:
> You'd get good test coverage, too; we'd be able to use the new build() in elem_refinement.C, MeshTools::Modification::flatten, and UnstructuredMesh::copy_nodes_and_elements().

**Comment by roystgnr** [comment]
I think John's concern is with the term "copy constructor".  In the literal C++ meaning, I'm not sure an Elem copy constructor makes sense either; if we added one we'd really never use it except in the new proposed Elem::build().

I think the term of art for the new Elem::build() overload I suggested is "virtual copy constructor idiom"?

Basically, instead of forcing the above user codes (as well as a new app code @loganharbour is working on) to ```build(old_elem->type())``` and then copy unique_id and elem integers and subdomain id data etcetera by hand (which is hard to get right with existing data and impossible to get future-proofed if we ever add new data like the elem integers again), we'd have them ```build(old_elem)```, which would do all that copying internal to libMesh, where we could change it in one place instead of five if it ever needs further changing.

**Comment by roystgnr** [comment]
My preferred suggestion was going to take an elem (to copy from) as well as an optional type (to allow first->second and second->first and tensor->simplex conversions to use the API too) ... except that wow, that really would be impossible not to get confused with the existing Elem::build arguments of a type and an optional elem (to be a *parent* of the new element...).  So I guess we'd want to have a new name for this, not just a new overload.

**Comment by loganharbour** [comment]
> So I guess we'd want to have a new name for this, not just a new overload.

`Elem::build_copy(const Elem *, const ElemType)` ?

**Comment by jwpeterson** [comment]
> I think the term of art for the new Elem::build() overload I suggested is "virtual copy constructor idiom"?

I see. Elsewhere in the library, we do this with `clone()` methods (which would call copy constructor internally), but I guess you are after a static function similar to the current `Elem::build()`? To me, if you are copying something, it makes sense to `clone()` the object which you have rather than calling a static function.

**Comment by roystgnr** [comment]
Hmm... should we just use clone(), like most of our other classes using this idiom?  And make it a member function, since it'll only make sense when we have an Elem to clone to call it on?  And then the default ElemType could be INVALID_ELEM, as an indicator to mean "just use this->type()".

**Comment by roystgnr** [comment]
John beat me by like 5 seconds.

**Comment by jwpeterson** [comment]
:laughing: let the record show that it was about 20 seconds

**Comment by loganharbour** [comment]
Ha - he's on top of his game. 

`Elem::clone(const ElemType = INVALID_ELEM` works for me.

**Comment by jwpeterson** [comment]
Sorry, what is the default for in the `clone()` case?

**Comment by jwpeterson** [comment]
Oh, OK, now that I understand the use case better, it seems like `clone()` is not really what we want... that just implies a duplicate of the original IMO.

**Comment by loganharbour** [comment]
For context: the use case that started this conversation is the action of copying an element to be inserted into another mesh.

**Comment by jwpeterson** [comment]
> For context: the use case that started this conversation is the action of copying an element to be inserted into another mesh.

OK, in that case calling `clone()` feels like the right thing to do. There's a bit of an issue regarding the Node and parent/interior_parent pointers in this case. Basically you don't want to shallow copy the original Elem's pointers into the clone in this case, instead they need to point to stuff in the other Mesh. This is why `UnstructuredMesh::copy_nodes_and_elements` and all of its complexity exists.

**Comment by loganharbour** [comment]
>  There's a bit of an issue regarding the Node and parent/interior_parent pointers in this case. 

Yeah, @roystgnr  and I had this discussion. The thought was to keep all of the nodes as `nullptr`. That way, it can be used for a variety of situations. And, if a user forgets to set nodes, the error of a null node will be a lot more useful than trying to set them behind the scenes.

**Comment by jwpeterson** [comment]
OK, I would suggest something like `skeleton_clone()` or some other name that makes it clear you aren't getting back a fully fleshed-out object if you decide to go this route. When I see `clone()` in practice, my assumption is that I can begin to use that object immediately as-is.
